### ISSUE TITLE
"Couldn't move remaining userspace processes, ignoring: Input/output error" message on dmesg/journalctl on boot since Linux 5.5 upgrade

### ISSUE THREAD

**Issue by joanbm** [systemd/systemd#14682]
Since upgrading my Arch Linux install to Linux 5.5, I get a "Couldn't move remaining userspace processes, ignoring: Input/output error" message on dmesg/journalctl on every boot. I have not observed any ill effect from it. I have found other closed issues with this message but they do not seem related to this instance.

I have successfully reproduced the error message on a clean Arch Linux install on a VM, using the Linux 5.5 package from the testing repositories. The message doesn't reproduce with Linux 5.4.15 from the official repositories, and also happened with Linux 5.5-rc7 from the mainline branch (i.e. without Arch-specific patches).

Here's the relevant excerpt from dmesg:

```
[    2.886708] systemd[1]: systemd 244.1-1-arch running in system mode. (+PAM +AUDIT -SELINUX -IMA -APPARMOR +SMACK -SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD +IDN2 -IDN +PCRE2 default-hierarchy=hybrid)
[    2.902930] systemd[1]: Detected architecture x86-64.
[    2.918709] systemd[1]: Set hostname to <SOL>.
[    2.920478] systemd[1]: Couldn't move remaining userspace processes, ignoring: Input/output error
[    3.064976] systemd[1]: Created slice system-getty.slice.
[    3.065240] systemd[1]: Created slice system-systemd\x2dcryptsetup.slice.
[    3.065451] systemd[1]: Created slice User and Session Slice.
```

**Comment by poettering** [comment]
hmm, there appear to be some processes we cannot move between cgroups during boot. Any chance you can boot with init=/bin/sh and then see if you can track them down with "ps xauwf"?

**Comment by joanbm** [comment]
I am not sure what I should exactly look after after booting with `init=/bin/sh` and printing the process list. I didn't see anything overly weird in it, just my shell and other kernel processes.

Other than that, I have tried to do some printf debugging and I have some results. The input/output error originates from `cgroup-util.c --> cg_read_pid` and propagates back to `cgroup-setup.c --> cg_migrate` and finally back to `cgroup.c --> manager_setup_cgroup` where the message is printed.

The file that `cg_read_pid` is reading is `/sys/fs/cgroup/unified/cgroup.procs` and the EIO is returned because the `fscanf` parses an unexpected value:

```
        errno = 0;
        if (fscanf(f, "%lu", &ul) != 1) {

                if (feof(f))
                        return 0;

                return errno_or_else(EIO);
        }

        if (ul <= 0)
                return -EIO; // <-- HERE!!!
```

Now if I `cat /sys/fs/cgroup/unified/cgroup.procs` after the system fully boots I can certainly see `0` values there:

```
2
3
4
5
6
7
8
9
10
11
12
13
14
0
0
0
16
17
18
[...]
```

It certainly seems unexpected given the above code but I'm not exactly sure what could be the cause of this or if I've just gone astray somewhere.

**Comment by poettering** [comment]
Hmm, looks like a kernel bug. It shows processes with PID 0 as cgroups? that cannot be right. @htejun ?

**Comment by joanbm** [comment]
I've done a bisect on the kernel and finally reached commit 1b96a41b420972f8ba563373de8154d59d5b2208 ( https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=1b96a41b420972f8ba563373de8154d59d5b2208 ).

**Comment by joanbm** [comment]
After more bisects I've reached 5153faac18d293fc7abb19ff7034683fbcd82dc7 ( https://git.kernel.org/pub/scm/linux/kernel/git/tj/cgroup.git/commit/?h=for-5.5&id=5153faac18d293fc7abb19ff7034683fbcd82dc7 ). Reverting this fixes the warning and makes the 0s in `/sys/fs/cgroup/unified/cgroup.procs` disappear.

**Comment by joanbm** [comment]
UPDATE: I can also see the 0's in `cgroup.procs` if I boot with `init=/bin/bash` and then mount with `mount -t cgroup2 none /mnt`. AFAIK no systemd code has run at that point.

**Comment by joanbm** [comment]
I've created a small init to try to reproduce the issue by printing `cgroup.procs`:

```
#include <sys/mount.h>
#include <sys/reboot.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <stdio.h>
#include <unistd.h>
#include <errno.h>

int main(void) {
	int r = mkdir("/mnt", 0755);
	if (r != 0) {
		printf("mkdir failed: %i\n", errno);
		reboot(RB_POWER_OFF);
	}

	r = mount("none", "/mnt", "cgroup2", 0, NULL);
	if (r != 0) {
		printf("mount failed: %i\n", errno);
		reboot(RB_POWER_OFF);
	}

	FILE *f = fopen("/mnt/cgroup.procs", "r");
	if (f == NULL) {
		printf("fopen failed: %i\n", errno);
		reboot(RB_POWER_OFF);
	}

	int c;
	while ((c = fgetc(f)) != EOF)
		putchar(c);

	printf("Done\n");
	reboot(RB_POWER_OFF);
}
```

Then I created a small initrd: `cc -static -o init init.c && echo init | cpio -H newc -o | gzip > initrd.gz`

Now, here's the interesting thing:

* If i boot with `qemu-system-x86_64 -nographic -append console=ttyS0 -kernel /boot/vmlinuz-linux  -initrd initrd.gz`, then no zeros appear

* But if I add `-smp 10` to the previous command line, then 9 zeros appear! In general, N-1 zeros are printed where N=number of SMPs

Unless somehow those zeros in `cgroup.procs` are expected behaviour (found no mention of it), then this is a simple test case where they appear without systemd.

**Comment by htejun** [comment]
After my patch, it was incorrectly adding init_tasks to root css_set. Will apply the fix right away and tag for -stable backport. Thank you so much for the report and bisecting.

**Comment by joanbm** [comment]
@htejun Nice, thanks for the confirmation and good to know a fix is in the works :)

Since this is a kernel issue I think this bug can be closed.

**Comment by htejun** [comment]
http://lkml.kernel.org/r/20200130164717.GE180576@mtj.thefacebook.com

**Comment by joanbm** [comment]
> http://lkml.kernel.org/r/20200130164717.GE180576@mtj.thefacebook.com

Thanks! I've applied the patch to my machine and so far everything is working right, the dmesg message is gone and no zeros in `cgroup.procs`.

**Comment by freefreeno** [comment]
Just to add that I am seeing this with kernel 5.5-1 in Arch.
Couldn't move remaining userspace processes, ignoring: Input/output error

**Comment by freefreeno** [comment]
5.5-2 Arch Linux
[josh@archlinux ~]$ dmesg --level=err,warn
[    0.464775] MDS CPU bug present and SMT on, data leak possible. See https://www.kernel.org/doc/html/latest/admin-guide/hw-vuln/mds.html for more details.
[    0.504228]   #7  #8  #9 #10 #11
[    0.668541] ENERGY_PERF_BIAS: Set to 'normal', was 'performance'
[    1.205125] hpet_acpi_add: no address or irqs in _CRS
[    1.422422] nvme nvme0: missing or invalid SUBNQN field.
[    2.140106] [drm:lspcon_init [i915]] *ERROR* Failed to probe lspcon
[    2.140135] [drm:intel_ddi_init [i915]] *ERROR* LSPCON init failed on port D
[    2.155364] [Firmware Bug]: ACPI(PEGP) defines _DOD but not _DOS
[    2.568892] i8042: PNP: PS/2 appears to have AUX port disabled, if this is incorrect please boot with i8042.nopnp
[    2.571773] i8042: Warning: Keylock active
[    2.587868] usb: port power management may be unreliable
[    2.739082] systemd[1]: Couldn't move remaining userspace processes, ignoring: Input/output error
[    3.220061] acpi PNP0C14:02: duplicate WMI GUID 05901221-D566-11D1-B2F0-00A0C9062910 (first instance was on PNP0C14:01)
[    3.220127] wmi_bus wmi_bus-PNP0C14:03: WQBC data block query control method not found
[    3.220129] acpi PNP0C14:03: duplicate WMI GUID 05901221-D566-11D1-B2F0-00A0C9062910 (first instance was on PNP0C14:01)
[    3.249242] acpi PNP0C14:04: duplicate WMI GUID 05901221-D566-11D1-B2F0-00A0C9062910 (first instance was on PNP0C14:01)
[    3.300161] acpi PNP0C14:05: duplicate WMI GUID 05901221-D566-11D1-B2F0-00A0C9062910 (first instance was on PNP0C14:01)
[    3.500267] i801_smbus 0000:00:1f.4: Accelerometer lis3lv02d is present on SMBus but its address is unknown, skipping registration
[    3.614888] iwlwifi 0000:6e:00.0: Direct firmware load for iwlwifi-cc-a0-52.ucode failed with error -2
[    3.614898] iwlwifi 0000:6e:00.0: Direct firmware load for iwlwifi-cc-a0-51.ucode failed with error -2
[    3.656298] i2c_hid i2c-DELL0926:00: i2c-DELL0926:00 supply vdd not found, using dummy regulator
[    3.656306] i2c_hid i2c-DELL0926:00: i2c-DELL0926:00 supply vddl not found, using dummy regulator
[    3.713802] ATPX version 1, functions 0x00000033
[    3.733235] CRAT table not found
[    3.812989] kvm: disabled by bios
[    4.029471] kvm: disabled by bios
[    4.100490] kvm: disabled by bios
[    4.200870] kvm: disabled by bios
[    4.276465] kvm: disabled by bios
[    4.382621] kvm: disabled by bios
[    4.471965] thermal thermal_zone10: failed to read out thermal zone (-61)
[    4.502548] kvm: disabled by bios
[    4.607571] kvm: disabled by bios
[    4.721098] kvm: disabled by bios
[    4.821333] kvm: disabled by bios
[    4.901557] iwlwifi 0000:6e:00.0: FW already configured (0) - re-configuring
[    4.918367] kvm: disabled by bios
[    4.918885] iwlwifi 0000:6e:00.0: BIOS contains WGDS but no WRDS
[    4.995823] kvm: disabled by bios
[    5.128636] iwlwifi 0000:6e:00.0: FW already configured (0) - re-configuring
[    5.144582] iwlwifi 0000:6e:00.0: BIOS contains WGDS but no WRDS
[    5.317708] [drm:dce110_edp_wait_for_hpd_ready [amdgpu]] *ERROR* dce110_edp_wait_for_hpd_ready: wait timed out!
[    5.824424] [drm:dce110_edp_wait_for_hpd_ready [amdgpu]] *ERROR* dce110_edp_wait_for_hpd_ready: wait timed out!
[    7.421920] kauditd_printk_skb: 18 callbacks suppressed
[   15.034392] kauditd_printk_skb: 2 callbacks suppressed
[   25.630001] iwlwifi 0000:6e:00.0: FW already configured (0) - re-configuring
[   25.639335] iwlwifi 0000:6e:00.0: BIOS contains WGDS but no WRDS
[   29.491441] [drm:dce110_edp_wait_for_hpd_ready [amdgpu]] *ERROR* dce110_edp_wait_for_hpd_ready: wait timed out!
[   29.991439] [drm:dce110_edp_wait_for_hpd_ready [amdgpu]] *ERROR* dce110_edp_wait_for_hpd_ready: wait timed out!
[   30.746031] kauditd_printk_skb: 3 callbacks suppressed

**Comment by sandy-8925** [comment]
Has this patch been applied to mainline yet? I'm still seeing this message on Linux 5.6-rc1.

**Comment by joanbm** [comment]
It hasn't been applied to any of the "bleeding edge" kernels (neither 5.5.3-rc2 nor 5.6-rc1).

**Comment by amartens1** [comment]
Hola Joan,

Is there a regression on 6.6.75.1 ? 
It looks the code has slightly changed and I am getting zeros again on `/sys/fs/cgroup/cgroup.procs` (I'm on cgroups V2 only) and the corresponding warning.

On `kernel/cgroup/cgroup.c`,  lines 6596+

```
        /* init tasks are special, only link regular threads */
        if (likely(child->pid)) {
                if (kargs->cgrp)
                        cgrp_flags = kargs->cgrp->flags;
                else
                        cgrp_flags = cset->dfl_cgrp->flags;

                WARN_ON_ONCE(!list_empty(&child->cg_list));
                cset->nr_tasks++;
                css_set_move_task(child, NULL, cset, false);
        } else {
                put_css_set(cset);
                cset = NULL;
        }
```

**Comment by amartens1** [comment]
@Werkov 

Thanks for the quick reply - indeed, the distro is the right place, but since Heo did the patch last time, and Joan filed the issue, I just wanted anybody else to confirm before opening the issue.

1) Not for the 6.6 branch, I've only tested 6.6.75.1. My latest 5.15 (5.15.167.4) does warn but contains zeros on `cgroup.procs`
2) Config attached
3) You are right, I have recompiled the kernel without the WARN_ONCE line and I still get the message. I'll have to dig a little bit more - any suggestion is welcomed 
 

[config.gz](https://github.com/user-attachments/files/18989133/config.gz)
